q2cli/builtin/tools.py

```diff
@@ -604,3 +604,247 @@ def citations(path):
+                              readable=True),
+              help='Path to an existing cache to remove the key from.')
+@click.option('--key', required=True,
+              help='The key to remove from the cache.')
+def cache_remove(path, key):
+    from qiime2.core.cache import Cache
+    from q2cli.core.config import CONFIG
+
+    try:
+        cache = Cache(path)
+        cache.remove(key)
+    except Exception as e:
+        header = "There was a problem removing the key '%s' from the " \
+                 "cache '%s':" % (key, path)
+        q2cli.util.exit_with_error(e, header=header, traceback=None)
+
+    success = "Removed key '%s' from cache '%s'" % (key, path)
+    click.echo(CONFIG.cfg_style('success', success))
+
+
+@tools.command(name='cache-garbage-collection',
+               short_help='Runs garbage collection on the cache at the '
+                          'specified location.',
+               help='Runs garbage collection on the cache at the specified '
+                    'location if the specified location is a cache.',
+               cls=ToolCommand)
+@click.option('--path', required=True,
+              type=click.Path(exists=True, file_okay=False, dir_okay=True,
+                              readable=True),
+              help='Path to an existing cache to run garbage collection on.')
+def cache_garbage_collection(path):
+    from qiime2.core.cache import Cache
+    from q2cli.core.config import CONFIG
+
+    try:
+        cache = Cache(path)
+        cache.garbage_collection()
+    except Exception as e:
+        header = "There was a problem running garbage collection on the " \
+            "cache at '%s':" % path
+        q2cli.util.exit_with_error(e, header=header, traceback=None)
+
+    success = "Ran garbage collection on cache at '%s'" % path
+    click.echo(CONFIG.cfg_style('success', success))
+
+
+@tools.command(name='cache-save',
+               short_help='Saves a .qza into the cache under a key.',
+               help='Saves a .qza into the cache under a key.',
+               cls=ToolCommand)
+@click.option('--cache-path', required=True,
+              type=click.Path(exists=True, file_okay=False, dir_okay=True,
+                              readable=True),
+              help='Path to an existing cache to save into.')
+@click.option('--artifact-path', required=True,
+              type=click.Path(exists=True, file_okay=True, dir_okay=False,
+                              readable=True),
+              help='Path to a .qza to save into the cache.')
+@click.option('--key', required=True,
+              help='The key to save the artifact under (must be a valid '
+                   'Python identifier).')
+def cache_save(cache_path, artifact_path, key):
+    from qiime2 import Artifact
+    from qiime2.core.cache import Cache
+    from q2cli.core.config import CONFIG
+
+    try:
+        artifact = Artifact.load(artifact_path)
+        cache = Cache(cache_path)
```

Comment:
If we `with` this cache before loading the artifact, then there will be fewer file operations.